### PR TITLE
compact and test

### PR DESCRIPTION
**Pull request by yanxiaoqi932**

### CODE REVIEW

File: vlog.go [hunk 1]
```diff
@@ -44,7 +48,7 @@ func openValueLog(options valueLogOptions) (*valueLog, error) {
 		vLogWal, err := wal.Open(wal.Options{
 			DirPath:        options.dirPath,
 			SegmentSize:    options.segmentSize,
-			SegmentFileExt: fmt.Sprintf(valueLogFileExt, i),
+			SegmentFileExt: fmt.Sprintf(valueLogFileExt, time.Now().Format("02-03-04-05-2006"), i),
```
Comment (roseduan) [inline vlog.go:51]: Can we add a suffix like `.temp`, after compaction, we can erase the suffix name to make it as a normal WAL file.

File: vlog.go [hunk 2]
```diff
@@ -35,6 +36,9 @@ type valueLogOptions struct {
 
 	// hash function for sharding
 	hashKeyFunction func([]byte) uint64
+
+	// writing validEntries to disk after reading the specified number of entries.
+	numEntriesToCompact int
```
Comment (roseduan) [inline vlog.go:41]: Name `CompactBatchCount` may be more suitable.

File: db_test.go [hunk 3]
```diff
@@ -385,6 +387,89 @@ func TestDBFlushMemTables(t *testing.T) {
...
+		for _, log := range testlogs {
+			value, err := getValueFromVlog(db, log.key)
+			assert.Nil(t, err)
+			assert.Equal(t, log.value, value)
+		}
+	})
+
+}
+
+func dirSize(path string) (int64, error) {
```
Comment (roseduan) [inline db_test.go:462]: add it in utils

File: db.go [hunk 4]
```diff
@@ -323,8 +337,110 @@ func (db *DB) flushMemtables() {
...
+			db.flushLock.Unlock()
+			return err
+		}
+	}
+
+	db.flushLock.Unlock()
+	return nil
+}
+
+func (db *DB) writeCompaction(newWal *wal.WAL, validEntries []*ValueLogRecord, part int) error {
```
Comment (roseduan) [inline db.go:431]: writeCompactionEntries

File: db.go [hunk 5]
```diff
@@ -323,8 +337,110 @@ func (db *DB) flushMemtables() {
...
+			newVLogWal, err := wal.Open(wal.Options{
+				DirPath:        db.vlog.options.dirPath,
+				SegmentSize:    db.vlog.options.segmentSize,
+				SegmentFileExt: fmt.Sprintf(valueLogFileExt, time.Now().Format("02-03-04-05-2006"), part),
+				BlockCache:     db.vlog.options.blockCache,
+				Sync:           false, // we will sync manually
+				BytesPerSync:   0,     // the same as Sync
+			})
+			if err != nil {
+				newVLogWal.Delete()
```
Comment (roseduan) [inline db.go:364]: handle the error returned from Delete

File: db.go [hunk 6]
```diff
@@ -323,8 +337,110 @@ func (db *DB) flushMemtables() {
...
 }
+
+func (db *DB) compact() error {
+	db.flushLock.Lock()
+	groups := make([]errgroup.Group, db.vlog.options.partitionNum)
+
+	for i := 0; i < int(db.vlog.options.partitionNum); i++ {
+		part := i
+		groups[part].Go(func() error {
+			newVLogWal, err := wal.Open(wal.Options{
```
Comment (roseduan) [inline db.go:355]: newValueFiles or tempValueFiles

File: vlog.go (hunk 1 above)
Comment (yanxiaoqi932) [reply vlog.go:51]: > Can we add a suffix like `.temp`, after compaction, we can erase the suffix name to make it as a normal WAL file.

I have considered this method, but in this way, maybe we need to add a rename fnnction in WAL. If we use time stamp to build wal file, there are not need to rename after compaction.

**Pull request merged**